catalyst-toolbox-lib/src/recovery/tally.rs

```diff
@@ -0,0 +1,648 @@
+        let block0_configuration = Block0Configuration::from_block(&block0)?;
+        let (_, voteplan) = voteplans.iter().next().ok_or(Error::MissingVoteplanError)?;
+        let vote_start = voteplan.vote_start();
+        let vote_end = voteplan.vote_end();
+        let committee_end = voteplan.committee_end();
+        let block0_start = block0_configuration.blockchain_configuration.block0_date;
+        let slot_duration = block0_configuration.blockchain_configuration.slot_duration;
+        let timeframe = timeframe_from_block0_start_and_slot_duration(block0_start, slot_duration);
+        let era = TimeEra::new(
+            Slot::from(0),
+            Epoch(0),
+            block0_configuration
+                .blockchain_configuration
+                .slots_per_epoch
+                .into(),
+        );
+        Ok(Self {
+            block0: block0.header.hash().into(),
+            vote_start,
+            vote_end,
+            committee_end,
+            range_check,
+            timeframe,
+            era,
+            fragments,
+            replay_protection: HashSet::new(),
+            spending_counters: HashMap::new(),
+        })
+    }
+
+    fn validate_tx<P>(
+        &mut self,
+        identifier: &account::Identifier,
+        witness: &account::Witness,
+        transaction: &TransactionSlice<P>,
+    ) -> bool {
+        let spending_counter = self
+            .spending_counters
+            .entry(identifier.clone())
+            .or_default();
+        let (is_valid_tx, _) = verify_original_tx(
+            SpendingCounter::from(*spending_counter),
+            &self.block0.into_hash(),
+            &transaction.transaction_sign_data_hash(),
+            identifier,
+            witness,
+            self.range_check.clone(),
+        );
+        is_valid_tx
+    }
+
+    fn validate_timestamp(
+        &self,
+        timestamp: SecondsSinceUnixEpoch,
+        start: &BlockDate,
+        end: &BlockDate,
+    ) -> bool {
+        fragment_log_timestamp_to_blockdate(timestamp, &self.timeframe, &self.era)
+            .map_or(false, |ref blockdate| {
+                !(blockdate < start || blockdate >= end)
+            })
+    }
+}
+
+impl<I: Iterator<Item = PersistentFragmentLog>> Iterator for VoteFragmentFilter<I> {
+    type Item = Result<Fragment, (Fragment, ValidationError)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.fragments.next().and_then(|persistent_fragment_log| {
+            let PersistentFragmentLog { fragment, time } = persistent_fragment_log;
+            match fragment {
+                Fragment::VoteCast(ref transaction) => {
+                    let transaction_slice = transaction.as_slice();
+                    let is_valid_vote_cast = valid_vote_cast(&transaction_slice);
+                    if !is_valid_vote_cast {
+                        return Some(Err((fragment, ValidationError::InvalidVoteCast)));
+                    }
```

Comment:
If invalid fragments can get into the permanent log (IIRC the pre-mempool checks in the node are weaker), a single invalid fragment would break our tally. I think we should report invalid fragments and proceed.

Reply:
This error is reported later. It do not break but continues with the next one. Fragment in the error is just used for logging and attaching to unprocessed fragments report.

Reply:
I see, thank you.